Resolve conflict with JDK15's new `isEmpty` method on `CharSequence`

catap opened the pull request:
`java.lang.CharSequence` has `isEmpty` method since JDK 15 that makes impossibly to compile scala by JDK15.

Fix it on the way that allows to compiled with JDK15 and before JDK15.

Fixes https://github.com/scala/bug/issues/12172

catap commented:
This way allows only to compile but not use it.

Close pull request because it is a wrong and very bad way.

som-snytt commented:
OK I'm cancelling the 1-800-flowers. It's a very annoying issue, so the swing is appreciated.

catap commented:
@som-snytt I was very optimised about impact.

Right now I doubt that it can be solved.

SethTisue commented:
@catap Note that it would have been sufficient to PR the change on 2.11.x, unless notably different code was needed on the different branches. We merge everything forward from 2.11.x -> 2.12.x -> 2.13.x.

catap commented:
@SethTisue I can't just cherry-pick it because it created some conflicts => I solved it.

Anyway, after a day of thinking about this bug I see that it should be solved on two levels:
 - inside scala-library by my PRs
 - inside scalac by some hacks

So, I would like to reopen my PRs

catap commented:
@SethTisue I made the first attempt to force selection of scala `isEmpty`.

This allows to produce on JDK15 a bytecode that runs on JVM before 15 that is good.

Anyway, I'm a bit concern about my naive if, I feel that I also should add a check that both implicit came from `scala.Predef`.

I did some naive method but I afraid that it costs to much in performance.

If you have something nice in mind I'd love to ask some help.

dwijnand commented:
So this is the special-case-it-in-the-compiler way.  When I heard about this breakage I was thinking we could make the CharSequence extending implicit non-implicit (so binary compatible) and ~have another implicit class that brings CharSequence's methods without actually extending CharSequence~ ... except that's not forwards-compatible... 🤔

catap commented:
@dwijnand but by this "hack" scala produces the same bytecode that it produces on JDK before 15.

Other approach is broking a lot of compatible in strange places, and here I just help to make right decision :)

lrytz commented on src/compiler/scala/tools/nsc/typechecker/Implicits.scala:
```diff
@@ -447,12 +451,27 @@ trait Implicits {
       if (StatisticsStatics.areSomeColdStatsEnabled) statistics.incCounter(improvesCount)
       (info2 == NoImplicitInfo) ||
-      (info1 != NoImplicitInfo) && {
+      (info1 != NoImplicitInfo) && resolveJDK15CharSequenceConflict(info1, info2) || {
```
would be good for performance to do this only after checking the `improvesCache`

lrytz commented on src/compiler/scala/tools/nsc/typechecker/Implicits.scala:
```diff
@@ -447,12 +451,27 @@ trait Implicits {
+     * Force to select scala implementation of `isEmpty` from `charArrayOps`
+     * because it produces bytecode that can be run on JDK before 15
+     */
+    def resolveJDK15CharSequenceConflict(info1: ImplicitInfo, info2: ImplicitInfo) =
```
can we also limit this to the situation when we are searching for a conversion to `?{def isEmpty}`? This code lives within the class `class ImplicitSearch(val tree: Tree, val pt: Type, val isView: Boolean, val context0: Context, val pos0: Position = NoPosition, val isByNamePt: Boolean = false)`, so there's quite a bit of context around.

lrytz reviewed (commented):
I don't see any other way out than compiler support for `Array[Char].isEmpty`.

sjrd commented:
I just thought: what if we a (intrinsified) macro `def isEmpty: Boolean` directly in `class Array`, that would expand `x.isEmpty` into `x.length == 0`? As a macro, it wouldn't violate the binary compatibility requirements. And it might be cleaner than special-casing implicit conversion insertion.

lrytz commented:
@sjrd thanks for that idea, I'll give it a try. It sounds very doable and much less intrusive.

dwijnand commented:
Macro (or symbol table + backend) ideas should remember about Dotty.

SethTisue commented:
If we use a macro, does Dotty have a way to ignore/override it?

catap commented:
@SethTisue and macro _a bit_ different between 2.12, 2.13 and dotty...

catap committed:
Resolve conflict with JDK15's new `isEmpty` method on `CharSequence`

`java.lang.CharSequence` has `isEmpty` method since JDK 15 that makes impossibly
to compile scala by JDK15 because `mutable.StringBuilder` and `reflect.api.Names`
implements their own `isEmpty` and both of them inhers `java.lang.CharSequence`.

This commit fixed it.

Anyway, the issue much deeper than "can't compile" scala-library. Let assume a
code:
```
new Array[Char](0).isEmpty
```

This code can't be compiled on JDK15+ because `isEmpty` can be used from
`scala.collection.ArrayOps[Char]` and from `ArrayCharSequence` that inherits
`java.lang.CharSequence` with a brand new `isEmpty`.

As result scala compile can't build this code with error like
```
type mismatch;
 found   : Array[Char]
 required: ?{def isEmpty: ?}
Note that implicit conversions are not applicable because they are ambiguous:
 both method ArrayCharSequence in object Predef of type (arrayOfChars: Array[Char]): ArrayCharSequence
 and method charArrayOps in object Predef of type (xs: Array[Char]): scala.collection.ArrayOps[Char]
 are possible conversion functions from Array[Char] to ?{def isEmpty: ?}
```

This commit also introduce a special case to solve conflict between JDK15+ and
Scala `isEmpty` methods, and forces to seelct scala version because it produces
bytecode that can be run on JDK before 15.

Link to the issue: https://github.com/scala/bug/issues/12172

smarter commented:
If a macro is defined in Array.scala, dotty will have to add special-cases in the typer and in the scala 2 unpickler to handle it (just like it special-cases the s and f interpolators). This isn't a huge deal but it also means that any attempt at using an existing dotty release together with a new scala-library (which might come in from a dependency) will break when trying to call that method. That would be pretty annoying and worth avoiding I think.

lrytz commented:
But the special case in implicit search also has to be replicated in dotty, and it seems a lot more intrusive. I see that by adding `isEmpty` to `Array.scala` we break dotty immediately. A way to avoid this is to create that symbol only within the compiler, similar to [`Object.synchronized`](https://github.com/scala/scala/blob/7f6b956e1a49910abe6f47c2e6d2c690a8a4bdbc/src/reflect/scala/reflect/internal/Definitions.scala#L1256) (forced in [`init()`](https://github.com/scala/scala/blob/7f6b956e1a49910abe6f47c2e6d2c690a8a4bdbc/src/reflect/scala/reflect/internal/Definitions.scala#L1644)).

catap commented:
@lrytz I can easy backport this hack to dotty.

But before I do this, let make a decision about future of this way.

smarter commented:
> A way to avoid this is to create that symbol only within the compiler, similar to Object.synchronized (forced in init()).

The difference is that Object is always read from a Java classfile, whereas Array can be read from sources or from a classfile, so you'll need to special-case it in two places in the compiler, for dotty it's worse: we can either read from sources, from a Scala 2 pickled Array.class or from a Dotty pickled Array.class, so we'd have to special-case three places.

lrytz committed:
Add support for Array.isEmpty stub, remove special case in implicits

lrytz commented:
Pushed a first attempt at a syntehtic `Array.isEmpty` symbol, bootstrap failed

```
[error] scala.reflect.internal.SymbolTable.abort(SymbolTable.scala:28)
[error] scala.reflect.internal.Scopes$EmptyScope$.enterEntry(Scopes.scala:550)
[error] scala.reflect.internal.Scopes$Scope.enter(Scopes.scala:157)
[error] scala.reflect.internal.Definitions.scala$reflect$internal$Definitions$$enterNewNullaryMethod(Definitions.scala:49)
[error] scala.reflect.internal.Definitions$DefinitionsClass.Array_isEmpty$lzycompute(Definitions.scala:503)
```

Will take a look if I can get it to work, or have to add the stub method to Array.scala.

smarter commented:
> Will take a look if I can get it to work, or have to add the stub method to Array.scala.

I'd really like it if we could avoid that, otherwise I'd like to know asap if this will be done since it means we need to delay 3.0.0-M1 again to handle this.

dwijnand commented:
Personally I can live with JDK 15 support coming in 2.13.5 (i.e. in 3-4 months).

catap commented:
@lrytz do you see any advantages of macros? It looks more complex and more costly for dotty team.

smarter commented:
> Personally I can live with JDK 15 support coming in 2.13.5 (i.e. in 3-4 months).

That might be even worse since it means we could release a 3.0.0 before this is settled and so be incompatible with the next scala-library version.

dwijnand commented:
What I had in mind is that there would be another release of Scala 3 which also brings JDK 15 support.

smarter commented:
> What I had in mind is that there would be another release of Scala 3 which also brings JDK 15 support.

The thing I'm worried about is not JDK 15 support, it's that dotty depends on the Scala 2 standard library, by default this is whatever version we have in the build of dotty itself, but it could be evicted in your build if you have a dependency which depends on a newer version. If a macro method is added to Array.scala, the dotty compiler will try to load it (no matter what the JDK version is) and won't be able to cope with it, and so we'll end up with some Scala 3 releases being incompatible with some Scala 2 library releases, this would be super confusing and inconvenient.

dwijnand commented:
Agreed.  Perhaps sbt-dotty should pin the scala library version, similar to how sbt does out-the-box on `scalaVersion`?  That way you stay in control of what standard library is used and can ship the `isEmpty` fix when it's ready.

smarter commented:
I mean sure, but I have a thousand other things to worry about and all I'm asking is for a decision to be reached soon so I can work with that.

lrytz commented:
> Will take a look if I can get it to work, or have to add the stub method to Array.scala.

I don't think there's a clean way to get the synthetic method to work. In order for it to be visible it has to be forced. This works fine except when compiling the standard library using `-sourcepath`. In this case the `Array` symbol is has a `SourcefileLoader`, i.e., a lazy type which completes by sending the source file through the compiler pipeline.

Other synthetic symbols are forced when the `Global.Run` instance is created, in `definitions.init()`. Completing a `SourcefileLoader` this early doesn't work (`run.firstPhase` is still `null`). We'd have to find some other hook, ideally somehow add the method right after completing the `Array` symbol, but there's no good way to do that.

So to me it seems that the best solution is to add the stub to `Array.scala`.

lrytz commented:
Hmm, that doesn't work so easily either... After adding `isEmpty` to `Array.scala` the 2.13.3 compiler can no longer compile the standard library, because calls within the standard library (like `clazz.getDeclaredFields.isEmpty`) now resolve to the new method, but it's not rewritten by the 2.13.3 compiler. So the backend fails with `unexpected array call`.

lrytz commented:
My only idea here is to do it in 2 steps:

  1. Add compiler support for a stub `isEmpty` method in `Array.scala`
  2. After the next release when we update starr, actually add that method to `Array.scala`

This means it won't be fixed for 2.13.4, only for 2.13.5. We could use a mergely as starr to avoid that.

I think this is still better than special-casing implicit search.

catap commented:
@lrytz am I right that this way is broke compatibility with class file that was compiled with a new standard library and that is using on old one?

Let image that some library is updated a scala version that is used to build it to 2.13.5 and user of this library is using 2.13.3.

If I understand it right the end user will have very wired runtime exception.

lrytz commented:
The `Array.class` is never used at run-time, Scala arrays are encoded as native arrays in the JVM. A call `someArray.isEmpty` is translated to `someArray.length == 0`, which works also when an older Scala library is on the runtime classpath.

smarter commented:
> I think this is still better than special-casing implicit search.

This seems a lot more complicated then special-casing implicit search, all for a temporary solution anyway (since the correct solution is to change Predef.scala whenever we can break BC), we've also discussed adding a special case in implicit search like this in dotty at a recent dotty meeting and there wasn't any opposition to it, why do you think it's a worse solution?

lrytz commented:
It's not *that* temporary, we'd have to keep it forever in 2.13. Adding the method to `Array` preserves Scala semantics, so for example IntelliJ's typer would understand it. It's also nothing new conceptually, we do the same for the other `Array` methods.

smarter commented:
> so for example IntelliJ's typer would understand it.

That would need to be tested, I don't know how their typer deals with macro methods but I wouldn't be surprised if something breaks.

> After the next release when we update starr, actually add that method to Array.scala

Can already prepare a PR with that change so I can test dotty against the updated Array.scala?

smarter commented:
Also what prevents you from using an intermediate starr based on a nightly instead of waiting for ~3-4 months?

SethTisue commented:
re-STARRing on a mergely seems quite acceptable to me, we've done it a number of times before

lrytz commented:
> Also what prevents you from using an intermediate starr based on a nightly instead of waiting for ~3-4 months?

I mentioned that as an option above. We only do that when really necessary, I'd be fine doing it here.

I'm not sure if a nightly on our artifactory is OK or if we should publish something to maven central, for reproducability.

SethTisue commented:
Historically we have done it to Maven Central.

lrytz commented:
Testing the follow up after restarr here: https://github.com/scala/scala/pull/9288

smarter commented:
Another approach which could be considered would be to simply make ArrayCharSequence not implicit at the cost of breaking source-compat in hopefully rare cases (but not binary compat):
```diff
diff --git src/library/scala/Predef.scala src/library/scala/Predef.scala
index c9911f9299..de2b8984a8 100644
--- src/library/scala/Predef.scala
+++ src/library/scala/Predef.scala
@@ -388,12 +388,13 @@ object Predef extends LowPriorityImplicits {
   }

   /** @group implicit-classes-char */
-  implicit final class ArrayCharSequence(arrayOfChars: Array[Char]) extends CharSequence {
+  final class ArrayCharSequence(arrayOfChars: Array[Char]) extends CharSequence {
     def length: Int                                     = arrayOfChars.length
     def charAt(index: Int): Char                        = arrayOfChars(index)
     def subSequence(start: Int, end: Int): CharSequence = new runtime.ArrayCharSequence(arrayOfChars, start, end)
     override def toString                               = arrayOfChars.mkString
   }
+  def ArrayCharSequence(arrayOfChars: Array[Char]): ArrayCharSequence

   /** @group conversions-string */
   @inline implicit def augmentString(x: String): StringOps = new StringOps(x)
```

That's one less implicit conversion in the world which would be a good thing, and for people who really want a CharSequence backed by their Array[Char], the Internet tells me that can be easily accomplished using `java.nio.CharBuffer.wrap`. The transition can be made smoother by first deprecating ArrayCharSequence in 2.13.4 with a message mentioning the alternative, and then removing the implicit keyword in 2.13.5 (this is the approach I took in scalacheck for a similar issue: https://github.com/typelevel/scalacheck/pull/498)

dwijnand commented:
Yes!  So by giving it the same name it's not forward binary incompatible.  Of course... -.-

I'm happy with dropping source compatibility for implicitly enriching `Array[Char]`.

sjrd commented:
That does seem like a reasonable option, indeed.

lrytz commented:
Intriguing idea... Should we then also deprecate `SeqCharSequence`?

I looked a bit through the history
  - 34fe33a612 (2008, Scala 2.7) added `seqToCharSequence`, it seems there was no similar implicit conversion before
  - d5b02c8652 (2009, Scala 2.8) added `arrayToCharSequence`, probably for consistency
  - b5e9e4b950 optimized `subSequence` to avoid copying the array. The correspoinding ticket https://github.com/scala/bug/issues/5641 suggests that the conversion is used in parser combinators.

I guess the main question is how much code we'd break. The community build can tell something about that.

dwijnand commented:
Alt: https://github.com/scala/scala/pull/9292

dwijnand commented:
Went with #9292 instead.

Thank you, @catap, for jumping on this so quickly.  Although we didn't end up merging your fix, this was really helpful to think through different approaches. ❤️
